File: cpp/tests/source_mapping_tests.cpp
// Source-mapping contract of prism-runtime-segmentation-v1: every token
// and sentence of a raw-text analysis maps onto half-open UTF-8 byte
// ranges of the exact, unmodified input. The expected offsets are written
// as literals and shared with the Swift test suite, which pins byte-offset
// parity across the bindings.

#include "prism/segmentation.h"

#include <fstream>
#include <sstream>
#include <string>

#include <gtest/gtest.h>

namespace {

using prism::Utf8ByteRange;
using prism::segmentation::NorwegianPolicy;
using prism::segmentation::PretokenizedSentence;
using prism::segmentation::Segment;
using prism::segmentation::SegmentationPolicy;

using Ranges = std::vector<Utf8ByteRange>;

SegmentationPolicy TestPolicy()
{
    return SegmentationPolicy{{"f.eks."}, 8};
}

// Every boundary must lie on a UTF-8 codepoint boundary of the input.
bool OnCodepointBoundary(std::string_view text, std::size_t offset)
{
    if (offset > text.size()) {
        return false;
    }
    return offset == text.size()
        || (static_cast<unsigned char>(text[offset]) & 0xC0U) != 0x80U;
}

void ExpectValidMapping(std::string_view text, const std::vector<PretokenizedSentence>& sentences)
{
    for (const auto& sentence : sentences) {
        ASSERT_EQ(sentence.token_source_ranges.size(), sentence.tokens.size());
        ASSERT_FALSE(sentence.source_ranges.empty());
        std::size_t previous_end = 0;
        for (const auto& range : sentence.source_ranges) {
            EXPECT_LT(range.start, range.end);
            EXPECT_GE(range.start, previous_end);
            EXPECT_TRUE(OnCodepointBoundary(text, range.start));
            EXPECT_TRUE(OnCodepointBoundary(text, range.end));
            previous_end = range.end;
        }
        for (const auto& token_ranges : sentence.token_source_ranges) {
            ASSERT_FALSE(token_ranges.empty());
            std::size_t previous_token_end = 0;
            for (const auto& range : token_ranges) {
                EXPECT_LT(range.start, range.end);
                EXPECT_GE(range.start, previous_token_end);
                EXPECT_TRUE(OnCodepointBoundary(text, range.start));
                EXPECT_TRUE(OnCodepointBoundary(text, range.end));
                previous_token_end = range.end;
            }
        }
    }
}

TEST(SourceMapping, AsciiTokensMapToExactByteRanges)
{
    const std::string text = "Katten sov. Hunden sov.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 2u);
    ExpectValidMapping(text, sentences);

    const std::vector<std::vector<Ranges>> expected_tokens{
        {{{0, 6}}, {{7, 10}}, {{10, 11}}},
        {{{12, 18}}, {{19, 22}}, {{22, 23}}},
    };
    for (std::size_t sentence = 0; sentence < sentences.size(); ++sentence) {
        EXPECT_EQ(sentences[sentence].token_source_ranges, expected_tokens[sentence]);
    }
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{0, 11}}));
    EXPECT_EQ(sentences[1].source_ranges, (Ranges{{12, 23}}));
}

TEST(SourceMapping, LeadingAndTrailingWhitespaceShiftsRanges)
{
    const std::string text = "  Katten sov.  ";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].token_source_ranges,
        (std::vector<Ranges>{{{2, 8}}, {{9, 12}}, {{12, 13}}}));
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{2, 13}}));
}

TEST(SourceMapping, NorwegianMultibyteLettersCountBytes)
{
    // å and æ occupy two UTF-8 bytes each; the offsets count bytes, not
    // characters.
    const std::string text = "Blåbær smaker godt.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].tokens[0], "Blåbær");
    EXPECT_EQ(sentences[0].token_source_ranges,
        (std::vector<Ranges>{{{0, 8}}, {{9, 15}}, {{16, 20}}, {{20, 21}}}));
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{0, 21}}));
}

TEST(SourceMapping, ConsecutiveMultibyteLettersFormOneToken)
{
    const std::string text = "æøå er bokstaver.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].tokens[0], "æøå");
    EXPECT_EQ(sentences[0].token_source_ranges,
        (std::vector<Ranges>{{{0, 6}}, {{7, 9}}, {{10, 19}}, {{19, 20}}}));
}

TEST(SourceMapping, EmojiBeforeAndBetweenTokens)
{
    const std::string before = "🙂 Katten sov.";
    auto sentences = Segment(before, NorwegianPolicy());
    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(before, sentences);
    EXPECT_EQ(sentences[0].tokens[0], "🙂");
    EXPECT_EQ(sentences[0].token_source_ranges,
        (std::vector<Ranges>{{{0, 4}}, {{5, 11}}, {{12, 15}}, {{15, 16}}}));

    const std::string between = "Katten 🙂 sov.";
    sentences = Segment(between, NorwegianPolicy());
    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(between, sentences);
    EXPECT_EQ(sentences[0].token_source_ranges,
        (std::vector<Ranges>{{{0, 6}}, {{7, 11}}, {{12, 15}}, {{15, 16}}}));
}

TEST(SourceMapping, DecomposedCombiningMarkStaysOnCodepointBoundaries)
{
    // "a" plus combining ring (U+030A) is visually "å" but differently
    // encoded; the scanner works on codepoints, so the combining mark
    // becomes its own token and every boundary stays a codepoint boundary.
    const std::string text = "a\xCC\x8A er fin.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].token_source_ranges,
        (std::vector<Ranges>{{{0, 1}}, {{1, 3}}, {{4, 6}}, {{7, 10}}, {{10, 11}}}));
}

TEST(SourceMapping, RepeatedIdenticalTokensMapToDistinctOccurrences)
{
    // A find()-style reconstruction would collapse onto the first
    // occurrence; the carried mapping assigns each repetition its own bytes.
    const std::string text = "ja ja ja ja.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].token_source_ranges,
        (std::vector<Ranges>{{{0, 2}}, {{3, 5}}, {{6, 8}}, {{9, 11}}, {{11, 12}}}));
}

TEST(SourceMapping, RepeatedIdenticalSentencesMapToDistinctOccurrences)
{
    const std::string text = "Han sov. Han sov.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 2u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{0, 8}}));
    EXPECT_EQ(sentences[1].source_ranges, (Ranges{{9, 17}}));
    EXPECT_EQ(sentences[0].token_source_ranges[0], (Ranges{{0, 3}}));
    EXPECT_EQ(sentences[1].token_source_ranges[0], (Ranges{{9, 12}}));
}

TEST(SourceMapping, RestoredSentenceSpaceKeepsOriginalOffsets)
{
    // "veien.Et" misses a space; the repair splits the sentences without
    // shifting any offset: the restored space owns no input bytes.
    const std::string text = "Han går langs veien.Et sekund senere står han.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 2u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{0, 21}}));
    EXPECT_EQ(sentences[1].source_ranges, (Ranges{{21, 48}}));
    // "Et" starts directly after the period of the first sentence.
    EXPECT_EQ(sentences[1].tokens[0], "Et");
    EXPECT_EQ(sentences[1].token_source_ranges[0], (Ranges{{21, 23}}));
}

TEST(SourceMapping, MultipleRepairedBoundariesStayAligned)
{
    const std::string text = "De gikk.De kom.De sov.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 3u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{0, 8}}));
    EXPECT_EQ(sentences[1].source_ranges, (Ranges{{8, 15}}));
    EXPECT_EQ(sentences[2].source_ranges, (Ranges{{15, 22}}));
}

TEST(SourceMapping, WrappedLineMergesAcrossNewlineWhitespace)
{
    // The merged line wrap is whitespace in the original, so the sentence
    // stays one covering range and the tokens straddle the newline.
    const std::string text = "Katten\nhennes sov.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].tokens,
        (std::vector<std::string>{"Katten", "hennes", "sov", "."}));
    EXPECT_EQ(sentences[0].token_source_ranges,
        (std::vector<Ranges>{{{0, 6}}, {{7, 13}}, {{14, 17}}, {{17, 18}}}));
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{0, 18}}));
}

TEST(SourceMapping, CollapsedWhitespaceRunsKeepTokenOffsets)
{
    const std::string text = "Hun   sov.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].token_source_ranges,
        (std::vector<Ranges>{{{0, 3}}, {{6, 9}}, {{9, 10}}}));
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{0, 10}}));
}

TEST(SourceMapping, DehyphenatedLineWrapKeepsBothFragments)
{
    // The de-hyphenated model token stays "språkmodellen", but its source
    // mapping must point at the two contributing fragments — never at a
    // single invented range that would claim "-\n" as token content.
    const std::string text = "Dette er språk-\nmodellen til laget.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(text, sentences);
    ASSERT_EQ(sentences[0].tokens[2], "språkmodellen");
    EXPECT_EQ(sentences[0].token_source_ranges[2], (Ranges{{9, 15}, {17, 25}}));
    EXPECT_EQ(text.substr(9, 6), "språk");
    EXPECT_EQ(text.substr(17, 8), "modellen");
    // No token range may cover the removed "-\n".
    for (const auto& token_ranges : sentences[0].token_source_ranges) {
        for (const auto& range : token_ranges) {
            EXPECT_TRUE(range.end <= 15 || range.start >= 17);
        }
    }
    // The sentence splits at the removed hyphen instead of bridging it.
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{0, 15}, {17, 36}}));
}

TEST(SourceMapping, DottedAbbreviationStaysOneContiguousRange)
{
    const std::string text = "Vi har f.eks. kake.";
    const auto sentences = Segment(text, NorwegianPolicy());

    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(text, sentences);
    EXPECT_EQ(sentences[0].tokens[2], "f.eks.");
    EXPECT_EQ(sentences[0].token_source_ranges[2], (Ranges{{7, 13}}));
}

TEST(SourceMapping, UrlAndEmailTokensMapAsSingleRanges)
{
    const std::string url_text = "Se https://prism.no i dag.";
    auto sentences = Segment(url_text, NorwegianPolicy());
    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(url_text, sentences);
    EXPECT_EQ(sentences[0].tokens[1], "https://prism.no");
    EXPECT_EQ(sentences[0].token_source_ranges[1], (Ranges{{3, 19}}));

    const std::string email_text = "Skriv til post@prism.no i dag.";
    sentences = Segment(email_text, NorwegianPolicy());
    ASSERT_EQ(sentences.size(), 1u);
    ExpectValidMapping(email_text, sentences);
    EXPECT_EQ(sentences[0].tokens[2], "post@prism.no");
    EXPECT_EQ(sentences[0].token_source_ranges[2], (Ranges{{10, 23}}));
}

TEST(SourceMapping, ChunkingSlicesTokenRangesAndClipsSentenceRanges)
{
    std::string text;
    for (int index = 0; index < 19; ++index) {
        if (index > 0) {
            text += ' ';
        }
        text += "ord" + std::to_string(index);
    }
    text += '.';

    const auto sentences = Segment(text, TestPolicy());

    ASSERT_EQ(sentences.size(), 3u);
    ExpectValidMapping(text, sentences);
    // Chunk 2 starts at "ord8"; its ranges continue seamlessly where
    // chunk 1 ended, and each chunk's sentence range covers only its own
    // tokens.
    EXPECT_EQ(sentences[1].tokens[0], "ord8");
    EXPECT_EQ(sentences[1].token_source_ranges[0], (Ranges{{40, 44}}));
    EXPECT_EQ(sentences[0].source_ranges, (Ranges{{0, 39}}));
    EXPECT_EQ(sentences[1].source_ranges, (Ranges{{40, 85}}));
    EXPECT_EQ(sentences[2].source_ranges, (Ranges{{86, 104}}));
}

TEST(SourceMapping, ChunkHelperSlicesCallerProvidedRanges)
{
    PretokenizedSentence sentence;
    sentence.tokens = {"a", "b", "c"};
    sentence.has_space_before = {false, true, true};
    sentence.token_source_ranges = {{{0, 1}}, {{2, 3}}, {{4, 5}}};
    sentence.source_ranges = {{0, 5}};

    const auto chunks = prism::segmentation::Chunk(sentence, 2);

    ASSERT_EQ(chunks.size(), 2u);
    EXPECT_EQ(chunks[0].token_source_ranges,
        (std::vector<Ranges>{{{0, 1}}, {{2, 3}}}));
    EXPECT_EQ(chunks[0].source_ranges, (Ranges{{0, 3}}));
    EXPECT_EQ(chunks[1].token_source_ranges, (std::vector<Ranges>{{{4, 5}}}));
    EXPECT_EQ(chunks[1].source_ranges, (Ranges{{4, 5}}));
}

TEST(SourceMapping, EveryFixtureTokenStaysAnchored)
{
    for (const auto* fixture : {"skarvholmen-bokmaal", "fjellvatnet-nynorsk"}) {
        std::ifstream file(std::string(PRISM_REPOSITORY_ROOT) + "/data/examples/"
                + fixture + ".txt",
            std::ios::binary);
        ASSERT_TRUE(file) << "Checked-in fixture is missing.";
        std::stringstream buffer;
        buffer << file.rdbuf();
        const auto text = buffer.str();

        const auto sentences = Segment(text, NorwegianPolicy());
        ExpectValidMapping(text, sentences);

        // Document order: sentence ranges never move backwards.
        std::size_t previous_start = 0;
        for (const auto& sentence : sentences) {
            EXPECT_GE(sentence.source_ranges.front().start, previous_start);
            previous_start = sentence.source_ranges.front().start;
        }
    }
}

} // namespace

File: swift/Tests/PrismKitTests/SourceMappingTests.swift
import XCTest

@testable import PrismKit

/// Source-mapping contract of the runtime segmentation. The expected byte
/// offsets are the same literals as in the C++ suite
/// (`cpp/tests/source_mapping_tests.cpp`), which pins byte-offset parity
/// across the bindings.
final class SourceMappingTests: XCTestCase {
    private let policy = SegmentationPolicy(
        abbreviationTokens: ["f.eks."],
        maximumTokenCount: 8
    )

    private func expectValidMapping(
        _ text: String,
        _ sentences: [PretokenizedSentence],
        file: StaticString = #filePath,
        line: UInt = #line
    ) {
        let byteCount = text.utf8.count
        for sentence in sentences {
            XCTAssertEqual(
                sentence.tokenSourceRanges.count, sentence.tokens.count,
                file: file, line: line
            )
            XCTAssertFalse(sentence.sourceRanges.isEmpty, file: file, line: line)
            for range in sentence.sourceRanges + sentence.tokenSourceRanges.flatMap({ $0 }) {
                XCTAssertLessThan(range.start, range.end, file: file, line: line)
                XCTAssertLessThanOrEqual(range.end, byteCount, file: file, line: line)
                // Both boundaries must map back onto the original string —
                // the helper rejects mid-codepoint offsets.
                XCTAssertNotNil(range.range(in: text), file: file, line: line)
            }
        }
    }

    func testAsciiTokensMapToExactByteRanges() {
        let text = "Katten sov. Hunden sov."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 2)
        expectValidMapping(text, sentences)
        XCTAssertEqual(
            sentences[0].tokenSourceRanges,
            [
                [Utf8ByteRange(start: 0, end: 6)],
                [Utf8ByteRange(start: 7, end: 10)],
                [Utf8ByteRange(start: 10, end: 11)],
            ]
        )
        XCTAssertEqual(sentences[0].sourceRanges, [Utf8ByteRange(start: 0, end: 11)])
        XCTAssertEqual(sentences[1].sourceRanges, [Utf8ByteRange(start: 12, end: 23)])
    }

    func testNorwegianMultibyteLettersCountBytes() {
        let text = "Blåbær smaker godt."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 1)
        expectValidMapping(text, sentences)
        XCTAssertEqual(sentences[0].tokens[0], "Blåbær")
        XCTAssertEqual(
            sentences[0].tokenSourceRanges,
            [
                [Utf8ByteRange(start: 0, end: 8)],
                [Utf8ByteRange(start: 9, end: 15)],
                [Utf8ByteRange(start: 16, end: 20)],
                [Utf8ByteRange(start: 20, end: 21)],
            ]
        )
    }

    func testEmojiBeforeAndBetweenTokens() {
        let before = "🙂 Katten sov."
        var sentences = RuntimeSegmentation.segment(before, policy: .norwegian())
        XCTAssertEqual(sentences.count, 1)
        expectValidMapping(before, sentences)
        XCTAssertEqual(
            sentences[0].tokenSourceRanges,
            [
                [Utf8ByteRange(start: 0, end: 4)],
                [Utf8ByteRange(start: 5, end: 11)],
                [Utf8ByteRange(start: 12, end: 15)],
                [Utf8ByteRange(start: 15, end: 16)],
            ]
        )

        let between = "Katten 🙂 sov."
        sentences = RuntimeSegmentation.segment(between, policy: .norwegian())
        XCTAssertEqual(sentences.count, 1)
        expectValidMapping(between, sentences)
        XCTAssertEqual(
            sentences[0].tokenSourceRanges[1],
            [Utf8ByteRange(start: 7, end: 11)]
        )
    }

    func testDecomposedCombiningMarkStaysOnCodepointBoundaries() {
        // "a" plus combining ring (U+030A) is visually "å" but differently
        // encoded; every boundary stays a codepoint boundary of the input.
        let text = "a\u{030A} er fin."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 1)
        XCTAssertEqual(
            sentences[0].tokenSourceRanges,
            [
                [Utf8ByteRange(start: 0, end: 1)],
                [Utf8ByteRange(start: 1, end: 3)],
                [Utf8ByteRange(start: 4, end: 6)],
                [Utf8ByteRange(start: 7, end: 10)],
                [Utf8ByteRange(start: 10, end: 11)],
            ]
        )
    }

    func testRepeatedIdenticalTokensMapToDistinctOccurrences() {
        // A find()-style reconstruction would collapse onto the first
        // occurrence; the carried mapping assigns each repetition its bytes.
        let text = "ja ja ja ja."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 1)
        expectValidMapping(text, sentences)
        XCTAssertEqual(
            sentences[0].tokenSourceRanges,
            [
                [Utf8ByteRange(start: 0, end: 2)],
                [Utf8ByteRange(start: 3, end: 5)],
                [Utf8ByteRange(start: 6, end: 8)],
                [Utf8ByteRange(start: 9, end: 11)],
                [Utf8ByteRange(start: 11, end: 12)],
            ]
        )
    }

    func testRestoredSentenceSpaceKeepsOriginalOffsets() {
        let text = "Han går langs veien.Et sekund senere står han."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 2)
        expectValidMapping(text, sentences)
        XCTAssertEqual(sentences[0].sourceRanges, [Utf8ByteRange(start: 0, end: 21)])
        XCTAssertEqual(sentences[1].sourceRanges, [Utf8ByteRange(start: 21, end: 48)])
        XCTAssertEqual(sentences[1].tokens[0], "Et")
        XCTAssertEqual(
            sentences[1].tokenSourceRanges[0], [Utf8ByteRange(start: 21, end: 23)]
        )
    }

    func testDehyphenatedLineWrapKeepsBothFragments() {
        // The de-hyphenated model token stays "språkmodellen", but its
        // source mapping must point at the two contributing fragments —
        // never at a single invented range claiming "-\n" as token content.
        let text = "Dette er språk-\nmodellen til laget."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 1)
        expectValidMapping(text, sentences)
        XCTAssertEqual(sentences[0].tokens[2], "språkmodellen")
        XCTAssertEqual(
            sentences[0].tokenSourceRanges[2],
            [Utf8ByteRange(start: 9, end: 15), Utf8ByteRange(start: 17, end: 25)]
        )
        let fragments = sentences[0].tokenSourceRanges[2].map { fragment in
            String(text[fragment.range(in: text)!])
        }
        XCTAssertEqual(fragments, ["språk", "modellen"])
        for range in sentences[0].tokenSourceRanges.flatMap({ $0 }) {
            XCTAssertTrue(range.end <= 15 || range.start >= 17)
        }
        // The sentence splits at the removed hyphen instead of bridging it.
        XCTAssertEqual(
            sentences[0].sourceRanges,
            [Utf8ByteRange(start: 0, end: 15), Utf8ByteRange(start: 17, end: 36)]
        )
    }

    func testWrappedLineMergesAcrossNewlineWhitespace() {
        let text = "Katten\nhennes sov."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 1)
        expectValidMapping(text, sentences)
        XCTAssertEqual(sentences[0].sourceRanges, [Utf8ByteRange(start: 0, end: 18)])
    }

    func testCollapsedWhitespaceRunsKeepTokenOffsets() {
        let text = "Hun   sov."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 1)
        XCTAssertEqual(
            sentences[0].tokenSourceRanges,
            [
                [Utf8ByteRange(start: 0, end: 3)],
                [Utf8ByteRange(start: 6, end: 9)],
                [Utf8ByteRange(start: 9, end: 10)],
            ]
        )
    }

    func testAbbreviationUrlAndEmailStayContiguous() {
        var sentences = RuntimeSegmentation.segment("Vi har f.eks. kake.", policy: .norwegian())
        XCTAssertEqual(sentences[0].tokens[2], "f.eks.")
        XCTAssertEqual(
            sentences[0].tokenSourceRanges[2], [Utf8ByteRange(start: 7, end: 13)]
        )

        sentences = RuntimeSegmentation.segment("Se https://prism.no i dag.", policy: .norwegian())
        XCTAssertEqual(sentences[0].tokens[1], "https://prism.no")
        XCTAssertEqual(
            sentences[0].tokenSourceRanges[1], [Utf8ByteRange(start: 3, end: 19)]
        )

        sentences = RuntimeSegmentation.segment(
            "Skriv til post@prism.no i dag.", policy: .norwegian()
        )
        XCTAssertEqual(sentences[0].tokens[2], "post@prism.no")
        XCTAssertEqual(
            sentences[0].tokenSourceRanges[2], [Utf8ByteRange(start: 10, end: 23)]
        )
    }

    func testChunkingSlicesTokenRangesAndClipsSentenceRanges() {
        let words = (0..<19).map { "ord\($0)" }
        let text = words.joined(separator: " ") + "."

        let sentences = RuntimeSegmentation.segment(text, policy: policy)

        XCTAssertEqual(sentences.count, 3)
        expectValidMapping(text, sentences)
        XCTAssertEqual(sentences[1].tokens[0], "ord8")
        XCTAssertEqual(
            sentences[1].tokenSourceRanges[0], [Utf8ByteRange(start: 40, end: 44)]
        )
        XCTAssertEqual(sentences[0].sourceRanges, [Utf8ByteRange(start: 0, end: 39)])
        XCTAssertEqual(sentences[1].sourceRanges, [Utf8ByteRange(start: 40, end: 85)])
        XCTAssertEqual(sentences[2].sourceRanges, [Utf8ByteRange(start: 86, end: 104)])
    }

    func testRangeInStringMapsAndRejectsInvalidBounds() {
        let text = "🙂å ok"
        // å (2 bytes) follows the 4-byte emoji.
        let emoji = Utf8ByteRange(start: 0, end: 4)
        let aRing = Utf8ByteRange(start: 4, end: 6)
        XCTAssertEqual(String(text[emoji.range(in: text)!]), "🙂")
        XCTAssertEqual(String(text[aRing.range(in: text)!]), "å")

        // Mid-codepoint and out-of-bounds boundaries are rejected, not
        // rounded.
        XCTAssertNil(Utf8ByteRange(start: 1, end: 4).range(in: text))
        XCTAssertNil(Utf8ByteRange(start: 0, end: 5).range(in: text))
        XCTAssertNil(Utf8ByteRange(start: 0, end: 99).range(in: text))
    }

    func testLeadingAndTrailingWhitespaceShiftsRanges() {
        let text = "  Katten sov.  "
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 1)
        expectValidMapping(text, sentences)
        XCTAssertEqual(
            sentences[0].tokenSourceRanges,
            [
                [Utf8ByteRange(start: 2, end: 8)],
                [Utf8ByteRange(start: 9, end: 12)],
                [Utf8ByteRange(start: 12, end: 13)],
            ]
        )
        XCTAssertEqual(sentences[0].sourceRanges, [Utf8ByteRange(start: 2, end: 13)])
    }

    func testConsecutiveMultibyteLettersFormOneToken() {
        let text = "æøå er bokstaver."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 1)
        expectValidMapping(text, sentences)
        XCTAssertEqual(sentences[0].tokens[0], "æøå")
        XCTAssertEqual(
            sentences[0].tokenSourceRanges,
            [
                [Utf8ByteRange(start: 0, end: 6)],
                [Utf8ByteRange(start: 7, end: 9)],
                [Utf8ByteRange(start: 10, end: 19)],
                [Utf8ByteRange(start: 19, end: 20)],
            ]
        )
    }

    func testRepeatedIdenticalSentencesMapToDistinctOccurrences() {
        let text = "Han sov. Han sov."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 2)
        expectValidMapping(text, sentences)
        XCTAssertEqual(sentences[0].sourceRanges, [Utf8ByteRange(start: 0, end: 8)])
        XCTAssertEqual(sentences[1].sourceRanges, [Utf8ByteRange(start: 9, end: 17)])
        XCTAssertEqual(sentences[0].tokenSourceRanges[0], [Utf8ByteRange(start: 0, end: 3)])
        XCTAssertEqual(sentences[1].tokenSourceRanges[0], [Utf8ByteRange(start: 9, end: 12)])
    }

    func testMultipleRepairedBoundariesStayAligned() {
        let text = "De gikk.De kom.De sov."
        let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())

        XCTAssertEqual(sentences.count, 3)
        expectValidMapping(text, sentences)
        XCTAssertEqual(sentences[0].sourceRanges, [Utf8ByteRange(start: 0, end: 8)])
        XCTAssertEqual(sentences[1].sourceRanges, [Utf8ByteRange(start: 8, end: 15)])
        XCTAssertEqual(sentences[2].sourceRanges, [Utf8ByteRange(start: 15, end: 22)])
    }

    func testChunkHelperSlicesCallerProvidedRanges() {
        let sentence = PretokenizedSentence(
            tokens: ["a", "b", "c"],
            hasSpaceBefore: [false, true, true],
            tokenSourceRanges: [
                [Utf8ByteRange(start: 0, end: 1)],
                [Utf8ByteRange(start: 2, end: 3)],
                [Utf8ByteRange(start: 4, end: 5)],
            ],
            sourceRanges: [Utf8ByteRange(start: 0, end: 5)]
        )

        let chunks = RuntimeSegmentation.chunk(sentence, maximumTokenCount: 2)

        XCTAssertEqual(chunks.count, 2)
        XCTAssertEqual(
            chunks[0].tokenSourceRanges,
            [[Utf8ByteRange(start: 0, end: 1)], [Utf8ByteRange(start: 2, end: 3)]]
        )
        XCTAssertEqual(chunks[0].sourceRanges, [Utf8ByteRange(start: 0, end: 3)])
        XCTAssertEqual(chunks[1].tokenSourceRanges, [[Utf8ByteRange(start: 4, end: 5)]])
        XCTAssertEqual(chunks[1].sourceRanges, [Utf8ByteRange(start: 4, end: 5)])
    }

    func testEveryFixtureTokenStaysAnchored() throws {
        let repositoryRoot = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
        for fixture in ["skarvholmen-bokmaal", "fjellvatnet-nynorsk"] {
            let textURL = repositoryRoot.appendingPathComponent(
                "data/examples/\(fixture).txt"
            )
            let text = try String(contentsOf: textURL, encoding: .utf8)

            let sentences = RuntimeSegmentation.segment(text, policy: .norwegian())
            expectValidMapping(text, sentences)

            // Document order: sentence ranges never move backwards.
            var previousStart = 0
            for sentence in sentences {
                let start = try XCTUnwrap(sentence.sourceRanges.first).start
                XCTAssertGreaterThanOrEqual(start, previousStart, fixture)
                previousStart = start
            }
        }
    }
}
